sem/utils.py

```diff
@@ -481,3 +489,231 @@ def convert_environment_str_to_dict(log_components):
+            # level=function is different.
+            if groups[10] is None and groups[12] is None:
+                groups[10] = 'function'
+                groups[12] = ''
+
+            temp_dict = {
+                'Time': float(groups[1]),
+                'Context': groups[2],
+                'Component': groups[5],
+                'Function': groups[6],
+                'Arguments': groups[7],
+                'Level': groups[10],
+                'Message': groups[12]
+            }
+            log_list.append(temp_dict)
+
+    return log_list
+
+
+def insert_logs(logs, db):
+    """
+    Insert the logs in the tinydb instance passed.
+
+    Note: This function does not return anything.
+
+    Args:
+        logs (list): A list of logs to insert in database.
+        db (TinyDB instance): A TinyDB instace where the logs will be inserted.
+    """
+    if logs == [] or logs is None:
+        return
+
+    example_result = {
+        k: ['...'] for k in ['Time',
+                             'Context',
+                             'Component',
+                             'Function',
+                             'Arguments',
+                             'Level',
+                             'Message']
+    }
+
+    for log in logs:
+        # Verify log format is correct
+        # Only check the if the keys are consistent
+        if not(set(log.keys()) == set(example_result.keys())):
+            raise ValueError(
+                '%s:\nExpected: %s\nGot: %s' % (
+                    "Log dictionary does not correspond to database format",
+                    pformat(example_result, depth=2),
+                    pformat(log, depth=2)))
+
+    db.table('logs').insert_multiple(logs)
+
+
+def filter_logs(db,
+                context=[],
+                function=[],
+                time_begin=-1,
+                time_end=-1,
+                level=[],
+                components={}):
+    """
+    Filter the logs stored in the database.
+
+    Filters are applied on context, function name, log level and time.
+    Additionally the user can also filter each log component based on a
+    particular level using components dictionary.
+    For example, if the user specifies Context = [0, 1] and Function = [A, B]
+    the function will output logs in which (context == 0 or context == 1) and
+    (function == a or function == b).
+
+    Return a list of logs that satisfy all the passed filters.
+
+    Args:
+        db (TinyDB instance): A TinyDB instace where the logs are inserted.
+        context (list): A list of context based on which the logs will be
+            filtered.
+        function (list): A list of function names based on which the logs will
+            be filtered.
+        time_begin (float): Start timestamp (in seconds) of the time window.
+        time_end (float): End timestamp (in seconds) of the time window.
+        level (list): A list of log levels based on which the logs will be
```

Comment:
Terminology is important here: a _level_ is something like `level_debug` to identify the _severity classes_ `debug`, `warn`, and `error`. In this case, it doesn't make that much sense to allow for multiple levels, since they are typically used one at a time. 

Which is supported? Both? Best to state it clearly in the documentation!

Reply:
For now this supports only the severity classes. I mentioned this in the documentation.

Reply:
Fixed terminology in [236ab77](https://github.com/signetlabdei/sem/pull/51/commits/236ab77d9240bf286834426c28d522b2beca030e).